jxl/src/render/stages/noise.rs

```diff
@@ -0,0 +1,357 @@
+        )
+    }
+}
+
+impl RenderPipelineStage for ConvolveNoiseStage {
+    type Type = RenderPipelineInOutStage<f32, f32, 2, 2, 0, 0>;
+
+    fn uses_channel(&self, c: usize) -> bool {
+        c >= self.first_channel && c < self.first_channel + 3
+    }
+
+    fn process_row_chunk(
+        &mut self,
+        position: (usize, usize),
+        xsize: usize,
+        row: &mut [(&[&[f32]], &mut [&mut [f32]])],
+    ) {
+        for (input, output) in row {
+            for x in 0..xsize {
+                let mut others = 0.0;
+                for i in 0..5 {
+                    let offset = (x as i32 + i) as usize;
+                    others += input[0][offset];
+                    others += input[1][offset];
+                    others += input[3][offset];
+                    others += input[4][offset];
+                }
+                others += input[2][x];
+                others += input[2][x + 1];
+                others += input[2][x + 3];
+                others += input[2][x + 4];
+                output[0][x] = others * 0.16 + input[2][x + 2] * -3.84;
+            }
+        }
+    }
+}
+
+pub struct AddNoiseStage {
+    noise: Noise,
+    // TODO(zond): color_correlation
+    first_channel: usize,
+}
+
+impl AddNoiseStage {
+    pub fn new(noise: Noise, first_channel: usize) -> AddNoiseStage {
+        AddNoiseStage {
+            noise,
+            first_channel,
+        }
+    }
+}
+
+impl std::fmt::Display for AddNoiseStage {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "add noise for channels [{},{},{}]",
+            self.first_channel,
+            self.first_channel + 1,
+            self.first_channel + 2
+        )
+    }
+}
+
+struct StrengthEvalLut<'a> {
+    noise: &'a Noise,
+}
+
+impl StrengthEvalLut<'_> {
+    fn eval(&self, vx: f32) -> f32 {
+        let k_scale = (self.noise.lut.len() - 2) as f32;
+        let scaled_vx = f32::max(0.0, vx * k_scale);
+        let pre_floor_x = scaled_vx.floor();
+        let pre_frac_x = scaled_vx / pre_floor_x;
+        let floor_x = if scaled_vx >= k_scale + 1.0 {
+            k_scale
+        } else {
+            pre_floor_x
+        };
+        let frac_x = if scaled_vx >= k_scale + 1.0 {
+            1.0
+        } else {
+            pre_frac_x
+        };
+        let floor_x_int = floor_x as usize;
+        let low = self.noise.lut[floor_x_int];
+        let hi = self.noise.lut[floor_x_int + 1];
+        (hi - low) * frac_x + low
+    }
+}
+
+fn noise_strength(noise_model: &StrengthEvalLut, x: f32) -> f32 {
+    // clamp doesn't handle NaN the same way as min/max.
+    #[allow(clippy::manual_clamp)]
+    f32::max(0.0, f32::min(noise_model.eval(x), 1.0))
+}
+
+impl RenderPipelineStage for AddNoiseStage {
+    type Type = RenderPipelineInPlaceStage<f32>;
+
+    fn uses_channel(&self, c: usize) -> bool {
+        c >= self.first_channel && c < self.first_channel + 3
```

Comment:
This doesn't seem correct, we should be using the 3 color channels [0,3) and [first_noise_channel, first_noise_channel+3) 

Relatedly, since `process_row_chunk` only gets passed channels that are used, we should just use 0..3 and 3..6 there as indices.
Also, the constructor should probably assert `first_noise_channel >= 3`.

Reply:
Done